### DISCUSSION TITLE
Send carb values ​​from another app to Xdrip

### DISCUSSION THREAD

**Discussion by dinizmauricio** [NightscoutFoundation/xDrip#3128] [General]
I would like to send carb values ​​from another app to Xdrip. It would be a type of integration. When reading the documentation about webservice (https://github.com/NightscoutFoundation/xDrip/blob/master/Documentation/technical/Local_Web_Services.md#implementation), I was unable to confirm whether it is possible. Does anyone know if it's possible?

**Comment by swissalpine** [comment]
No, as far as I know this is not integrated yet. What is possible (e.g. steps, heart rate) is documented.

**Comment by dinizmauricio** [comment]
Could you tell me how I can propose this functionality/integration?

**Reply by emp-00** [reply]
You can create a fork and code the "integration" that you wish to have? The short description is not clear what exactly you're planning to do, at least to me.

**Comment by dinizmauricio** [comment]
@emp-00, from the client side, I would like to do this:

import requests
myHash='d018020ba8f49a010afdbabf7e8be89551085750'
header1 = {'api-secret': myHash}
new_data = {
    "carbs": 10
}
resposta = requests.post('http://192.168.0.1:17580/entries', headers=header1, json=new_data)
print (resposta.text)

Do you think it is possible for the cell phone's webservice to receive this request?

**Reply by swissalpine** [reply]
Yes, this is certainly possible, but it should - presumably - be based on how data is passed to xDrip+ so far:
http://127.0.0.1:17580/sgv.json?steps=1234&heart=123
Have a look at the related documentation here:
https://github.com/NightscoutFoundation/xDrip/blob/master/Documentation/technical/Local_Web_Services.md

**Comment by dinizmauricio** [comment]
Yes, I had already seen the documentation and part of the code: https://github.com/NightscoutFoundation/xDrip/blob/c6b15c568a859106bbe3149cd544f391cb4ad203/app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java#L34


I think the amount of carbohydrates can be sent following the same pattern as other treatments: http://127.0.0.1:17580/sgv.json?steps=1234&heart=123&carb=80

But I confess that I don't have the competence to implement it on the webservice side and create a fork. If anyone can help....

**Comment by rickykenny** [comment]
I am also interested in passing carbs and treatments to xDrip from another app. I'm new to Android dev but very experienced in web dev, hopefully I can be useful 😅

In the interest of setting off in the right direction, I would like to first confirm that the web server is the best place to implement this?

I've found two other methods of passing data to xDrip:
1. [NSClientReceiver.java](https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/NSClientReceiver.java) - looks like this receives intents specifically from the NightScout client, so I think its not the right approach for us
2. [BroadcastService.java](https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/services/broadcastservice/BroadcastService.java) - looks like this receives broadcast intents and is enabled on the "inter-app settings" screen with the "Broadcast Service API" toggle? The caption says this "enables xDrip communication with third-party applications by using new API". This handles functions like `CMD_ADD_STEPS` and `CMD_ADD_TREATMENT` and the commit history suggests that [Watchdrip+](https://github.com/bigdigital/watchdrip) is using it:
https://github.com/NightscoutFoundation/xDrip/blob/c6b15c568a859106bbe3149cd544f391cb4ad203/app/src/main/java/com/eveningoutpost/dexdrip/services/broadcastservice/BroadcastService.java#L191

Personally, I'm leaning towards the Broadcast service because it requires apps to register their own identifier, so entries _can_ be linked to their source (although that's not yet implemented in xDrip). I also assume (naively) that enabling the web server would have a greater impact on battery life?

- Does the xDrip community have a preference for either method? Perhaps using the broadcast service would allow existing projects like Watchdrip+ to send carbs easier in the future.
- @dinizmauricio would this method be useful to you, or do you specifically need to use the webserver?

Also, can we please rename this discussion to "send carb values ​​from another app to Xdrip"?

**Comment by dinizmauricio** [comment]
I didn't know about the option to send via broadcasting and thank you for mentioning it.
In my case, it makes no difference whether to use broadcasting or webservice, the important thing is to send the carbohydrates from another application.

I already changed the name of the topic, it was a good suggestion.

**Comment by rickykenny** [comment]
Thanks @dinizmauricio. 
Looking further into the code, adding carbs seems to already be implemented under the `CMD_ADD_TREATMENT` intent:
https://github.com/NightscoutFoundation/xDrip/blob/c6b15c568a859106bbe3149cd544f391cb4ad203/app/src/main/java/com/eveningoutpost/dexdrip/services/broadcastservice/BroadcastService.java#L191-L195

I think this satisfies my requirements, now I just need to learn how to broadcast to this intent!

**Reply by dinizmauricio** [reply]
Dear @rickykenny 
Thanks for the information, I was researching [other discussions](https://github.com/NightscoutFoundation/xDrip/pull/2004)  and understood that the Xdrip can receive carbohydrates using broadcast, as you said. But I don't know where to start. If anyone can help create a small code that sends anything...

**Reply by rickykenny** [reply]
Thanks for the link to the previous conversation @dinizmauricio, that's super helpful 🙏
From what I can tell by looking at the WatchDrip+ code and that conversation, **(and this might be wrong...**) you need to:
1. Wait for xDrip to broadcast an intent with `FUNCTION` key set to `start`
1. register your app in xDrip by broadcasting an intent with the `FUNCTION` key set to `set_settings` and pass some setttings config (unsure what this is yet)
1. Now you can broadcast an intent with `FUNCTION` key set to `add_treatment` and, containing `carbs` or `insulin` values 

I'm writing a basic app in Flutter/Kotlin to learn how to do this - I'll post a link here if I manage to get it working.
What language are you writing your app in?

**Comment by dinizmauricio** [comment]
@rickykenny ,
I will try to develop using React Native.

I believe the beginning of the code is something like:

`import React, { useEffect } from 'react';
import { DeviceEventEmitter } from 'react-native';

export const App = () => {
  useEffect(() => {
    DeviceEventEmitter.addListener('broadcaster-data-received', (data) => {
      console.log('Your data is here: ', data);
    });
  }, []);
  return null;
};`
